File: RESTful service/audio.py
import sys
import os
import base64
import hmac
import hashlib
import urllib
import time
import requests

'''
This demo shows how to use the RESTful API to upload an audio file ( "data_type":"audio" ) into your bucket.
You can find account_access_key and account_access_secret in your account page.
Log into http://console.acrcloud.com -> "Your Name"(top right corner) -> "Account" -> "Console API Keys" -> "Create Key Pair". 
Be Careful, they are different with access_key and access_secret of your project.
'''
option = {
  'host': 'api.acrcloud.com',
  'signature_version': '1',
  'access_key': '<account_access_key>',
  'access_secret': '<account_access_scret>',
};


def sign(string_to_sign, access_secret):
    return  base64.b64encode(
		    hmac.new(access_secret.encode(), string_to_sign.encode(), digestmod=hashlib.sha1)
		    .digest())

def upload_audio(path, bucket, title, audio_id, data_type="audio", custom_fields=None):
    http_method = "POST"
    timestamp = str(time.time())
    http_uri = "/v1/audios"

    string_to_sign = '\n'.join((http_method, http_uri, option['access_key'], option['signature_version'], str(timestamp)))

    signature = sign(string_to_sign, option['access_secret'])

    f = open(path, "rb")

    files = {'audio_file':f}
    headers = {'access-key': option['access_key'], 'signature-version': option['signature_version'], 'signature': signature, 'timestamp':timestamp}
    #if you uplaod fingerprint file , please set "data_type":"fingerprint"
    data = {'title':title, "audio_id":audio_id, "bucket_name":bucket, "data_type":data_type}
    if custom_fields:
        keys = []
        values = []
        for k in custom_fields:
            keys.append(k)
            values.append(custom_fields[k])
        data['custom_key[]'] = keys
        data['custom_value[]'] = values

    requrl = "https://"+option['host'] + http_uri
    r = requests.post(requrl, files=files, data=data, headers=headers, verify=True)
    r.encoding = "utf-8"
    print r.text

def get_audios(bucket, page=1):
    http_method = "GET"
    timestamp = str(time.time())
    http_uri = "/v1/audios"

    string_to_sign = '\n'.join((http_method, http_uri, option['access_key'], option['signature_version'], str(timestamp)))
    signature = sign(string_to_sign, option['access_secret'])
    headers = {'access-key': option['access_key'], 'signature-version': option['signature_version'], 'signature': signature, 'timestamp':timestamp}

    requrl = "https://"+option['host'] + http_uri+"?bucket_name="+bucket+"&page="+str(page)
    r = requests.get(requrl, headers=headers, verify=True)
    r.encoding = "utf-8"
    print r.text

def update_audio(bucket, acr_id, title, audio_id, custom_fields=None):
    http_method = "PUT"
    timestamp = str(time.time())
    http_uri = "/v1/audios/"+acr_id

    string_to_sign = '\n'.join((http_method, http_uri, option['access_key'], option['signature_version'], str(timestamp)))
    signature = sign(string_to_sign, option['access_secret'])
    headers = {'access-key': option['access_key'], 'signature-version': option['signature_version'], 'signature': signature, 'timestamp':timestamp}

    data = {'title':title, "audio_id":audio_id, "bucket_name":bucket}
    if custom_fields:
        keys = []
        values = []
        for k in custom_fields:
            keys.append(k)
            values.append(custom_fields[k])
        data['custom_key[]'] = keys
        data['custom_value[]'] = values

    requrl = "https://"+option['host'] + http_uri
    r = requests.put(requrl, data=data, headers=headers, verify=True)
    r.encoding = "utf-8"
    print r.text

def delete_audio(acr_id):
    http_method = "DELETE"
    timestamp = str(time.time())
    http_uri = "/v1/audios/"+acr_id

    string_to_sign = '\n'.join((http_method, http_uri, option['access_key'], option['signature_version'], str(timestamp)))
    signature = sign(string_to_sign, option['access_secret'])
    headers = {'access-key': option['access_key'], 'signature-version': option['signature_version'], 'signature': signature, 'timestamp':timestamp}

    requrl = "https://"+option['host'] + http_uri
    r = requests.delete(requrl, headers=headers, verify=True)
    r.encoding = "utf-8"
    if r.status_code == 204:
        print "deleted"

if __name__ == "__main__":
    #update_audio('eu-test', 'db397154fb9858f17373b5bf66be875a', 'Hiding my heart - Adele', '12345', {"artist":"Adele", "title":"Hiding my heart"})
    #delete_audio("07221fadacc359fdb0744cd28b7b0ce9")
    #get_audios('eu-test', 1)
    upload_audio(sys.argv[1], 'eu-test', 'test-title', '1212', 'fingerprint')

File: RESTful service/audio.js
var url = require('url');
var path = require('path');
var fs = require('fs');
var crypto = require('crypto');
//npm install request
var request = require('request');

var defaultOptions = {
  host: 'api.acrcloud.com',
  endpoint: '/v1/audios',
  signature_version: '1',
  secure: true,
  access_key: 'xxx',
  access_secret: 'xxxxx'
};

function buildStringToSign(method, uri, accessKey, signatureVersion, timestamp) {
  return [method, uri, accessKey, signatureVersion, timestamp].join('\n');
}

function sign(signString, accessSecret) {
  return crypto.createHmac('sha1', accessSecret)
    .update(Buffer.from(signString, 'utf-8'))
    .digest().toString('base64');
}

/**
 * Identifies a sample of bytes
 */
function upload(file_path, bucket, title, audio_id, data_type="fingerprint", custom_fields=None, cb) {
  var current_data = new Date();
  var timestamp = current_data.getTime()/1000;

  var stringToSign = buildStringToSign('POST',
    defaultOptions.endpoint,
    defaultOptions.access_key,
    defaultOptions.signature_version,
    timestamp);
  console.log(stringToSign)

  var signature = sign(stringToSign, defaultOptions.access_secret);
  console.log(signature)

  var headers = {
    'access-key': defaultOptions.access_key,
    'signature-version': defaultOptions.signature_version,
    'signature': signature,
    'timestamp':timestamp
  };
  var formData = {
    'audio_file': {
        'value':  fs.createReadStream(file_path),
        'options': {
            'filename': path.basename(file_path),
            'contentType': 'fingerprint/lo'
        }
    },
    'data_type':data_type,
    'bucket_name':bucket,
    'title':title,
    'audio_id':audio_id,
  }
  if (custom_fields) {
        keys = []
        values = []
        for (var k in custom_fields) {
            keys.push(k)
            values.push(custom_fields[k])
        }
        formData['custom_key[]'] = keys
        formData['custom_value[]'] = values
  }

  request.post({
    url: "https://"+defaultOptions.host + defaultOptions.endpoint,
    method: 'POST',
    headers: headers,
    formData: formData
  }, cb);
}

function get_audio(acr_id, cb) {
  var current_data = new Date();
  var timestamp = current_data.getTime()/1000;

  var endpoint = defaultOptions.endpoint+"/"+acr_id

  var stringToSign = buildStringToSign('GET',
    endpoint,
    defaultOptions.access_key,
    defaultOptions.signature_version,
    timestamp);

  var signature = sign(stringToSign, defaultOptions.access_secret);

  var headers = {
    'access-key': defaultOptions.access_key,
    'signature-version': defaultOptions.signature_version,
    'signature': signature,
    'timestamp':timestamp
  };

  request.get({
    url: "https://"+defaultOptions.host + endpoint,
    method: 'GET',
    headers: headers,
  }, cb);
}

var title = "test"
var audio_id = "test"
var bucket_name = "eu-test"
var data_type = "fingerprint"
var file_path = '1040210008.mp3.wav.lo';
var custom_fields = {"key1":"value1"}

upload(file_path, bucket_name, title, audio_id, data_type, custom_fields , function (err, httpResponse, body) {
  if (err) console.log(err);
  console.log(body);
});

get_audio("4b616d1b695bf4c35ed5103f9cbdd507", function(err, httpResponse, body){
    console.log(body);
});
